rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2579 @@
+
+# Summary
+
+[summary]: #summary
+
+This RFC proposes changing the way data structural requirements are expressed within the BP.
+
+Rather than allowing any free-form JSON schema, this RFC defines a set of JSON schema meta-schemas, which describe:
+
+- Data Types
+- Property Types
+- Link Types
+- Entity Types
+
+These elements are used in combination to improve reusability of, and better define, expressions of data structural requirements.
+
+By using the proposed types, blocks, embedding applications, and end-users can all have greater confidence in each other's roles and functionalities, and hopefully will allow them all to benefit from better computational inference and capabilities. Reusability of descriptions of data should encourage a decrease in developmental overhead overtime, with all members of the ecosystem benefiting from emergent standardization.
+
+# Motivation
+
+[motivation]: #motivation
+
+The Block Protocol operates in-between parties that otherwise have limited to no communication. A block developer may not even be aware of the existence of an embedding application or the data within it, and in some scenarios the developers and owners of an embedding application may not be aware of the existence of a block (in the case of dynamic selection and loading at run-time). This is a common occurrence already in technology, whether it be library authors not knowing what projects are built with their work, or developers using libraries without utilizing pathways for communication (online forums or otherwise). This is made possible through encoded contracts: function signatures, documentation, REST APIs, etc. These tools are ways of asynchronously, and often autonomously, declaring one side of the communication channel. In such a system of such imperfect information, the standardization of communication mediums is essential to effective function.
+
+By getting both parties to agree to a prescriptive standard, new technological possibilities become viable. The Block Protocol is an example of this, enabling a new class of applications, ones that are able to dynamically load complex front-end components at run-time which are immediately usable and functional, even if the component was previously unknown to the application.
+
+The foundation of this capability is the standardization of the expression of requirements. Most notably, the block and entity schemas expressed through the graph service allow for dynamic resolution of expected input data.
+
+---
+
+**Example**:
+
+A block expresses that it needs a URL hosting an image (presumably so that it can render it):
+
+```json
+{
+  "properties": {
+    "image_url": "string"
+  }
+}
+```
+
+The embedding application is then able to search across its data to find entities that satisfy this requirement, such as:
+
+```json
+{
+  "image_url": "https://dummyimage.com/300.jpg",
+  ... // some other fields
+}
```

Comment:
@thehabbos007 and I were discussing this as well, we're not sure what the best solution is 🙃 
We tried `jsonc` but it actually removes a lot of the formatting. 

The big red boxes have grown on me if I'm honest, they kind of make it very clear that it's not part of the data, but I appreciate they could be very distracting. 

Are you able to attach a screenshot here showing the difference between JSON and JS formatting?

Reply:
I don't feel super strongly about this, but I do find the red boxes distracting. And I think it's better for the reader to be focusing on the actual code in the document, rather than our comments and `...` taking the centre stage.

But again it's just a preference.

Reply:
Fair enough, don't feel strongly so will resolve.